### PR TITLE
Update duration rounding to new algorithms

### LINKED ISSUE #19: Implement `Duration` normalization and new algorithms
The proposal recently updated `Duration` in tc39/proposal-temporal#2722, tc39/proposal-temporal#2727, and tc39/proposal-temporal#2612. This will need to be updated accordingly.

EDIT: There's been an overhaul of the implementation of the duration methods in general on top of normalization that needs to be updated.

### PR DESCRIPTION
**Pull request by nekevss**
~~Currently a WIP.~~

This closes #19 outside of anything related to ZonedDateTime, which I'd prefer to consider separately due to the reliance on time zones.

This PR updates duration rounding to the new duration rounding algorithms.

As it's a bit of a larger PR, I thought I would create a draft in case anyone wants to take a look early.

Overall, this draft is fairly close to complete, but there are still at least a couple bugs around the actually rounding that I'm working through.

### CODE REVIEW

**Comment by nekevss** [comment]
This PR should definitely wait until after #66 is merged so it can be rebased to remove the custom traits, but it is essentially ready for review outside of that impending rebase

There are a couple things to note:
  1. This implements the new rounding methods under a `round_v2` and doesn't remove the previous one just yet. It can be removed if everyone else thinks it should be in this PR, but I was sort of thinking about lagging on removing the previous methods while the rest is tested. Hopefully, it also makes the diff just a little bit easier to read. 
  2. Any rounding around `ZonedDateTime` is not included in this, but it's hard to implement and test that without proper time zone support, so if everyone else is in agreement I'd prefer to file that under a new issue.
  3. We should probably file an issue for more unit tests around the internal nudge algorithms and normalization records.

File: src/components/duration.rs
Comment (jedel1043) [inline src/components/duration.rs:461]: Seeing how many optionals are here, maybe we should extract this into a `RoundingOptions` struct that implements `Default`.

File: src/components/duration.rs
Comment (jedel1043) [inline src/components/duration.rs:456]: About the `round_v2`, I think it should just replace the old rounding function. We just need to test `Boa` against the PR branch to see that everything looks good.

File: src/components/duration/normalized.rs
Comment (jedel1043) [inline src/components/duration/normalized.rs:185]: Can we move this to the constructor?

File: src/components/duration/normalized.rs
Comment (jedel1043) [inline src/components/duration/normalized.rs:156]: I think it would be better to use a proper struct instead of a tuple here. Also, can we make this private? I'm not sure if we need to allow external mutations here, seeing that we require checks on `new` that may not be preserved by external code.

File: src/utils.rs
Comment (jedel1043) [inline src/utils.rs:23]: Seeing that this is only used by `sign`, maybe it can just be a wrapper around `NonZeroI8`?
```rust
struct Sign(NonZeroI8);

impl Sign {
    const POSITIVE: Self = const {
        let Some(n) = NonZeroI8::new(1) else {
            panic!("invalid number for sign");
        }
        Self(n)
    }

    const NEGATIVE: Self = const {
        let Some(n) = NonZeroI8::new(-1) else {
            panic!("invalid number for sign");
        }
        Self(n)
    }

    pub(crate) fn get(&self) -> i8 {
        self.0.get()
    }
}
```
This also makes `std::mem::size_of::<Option<Sign>>() == std::mem::size_of::<i8>()`

**Review by jedel1043 — COMMENTED** [review]
Looks great! I have some suggestions on some APIs that we can improve.

File: src/components/duration.rs
Comment (nekevss) [reply src/components/duration.rs:461]: Yeah, I was thinking about doing this in a follow up, but it could be added to this as well.

Temporal spec refers to them as difference settings `13.53 GetDifferenceSettings`, but I don't hate `RoundingOptions` as a name. Any preference between `DifferenceSettings` vs. `RoundingOptions`

Comment (jedel1043) [reply src/components/duration.rs:461]: I prefer `RoundingOptions` for consistency. If we need `DifferenceOptions` in some other place (the diff_instant operation maybe?), we can do a `type DifferenceOptions = RoundingOptions;` for it.

File: src/utils.rs
Comment (jedel1043) [reply src/utils.rs:23]: Actually, an `enum Sign` optimizes perfectly fine, apparently.
```rust
enum Sign {
    Positive = 1,
    Negative = -1,
}

dbg!(std::mem::size_of::<Sign>()); // 1
dbg!(std::mem::size_of::<Option<Sign>>()); // 1
```

Comment (nekevss) [reply src/utils.rs:23]: Hmmm maybe a Sign, which can be zero, and a NonZeroSign (hence the NonZeroOrdering).

Comment (jedel1043) [reply src/utils.rs:23]: Do we need a sign that can be zero though? I would expect it to only have Positive and Negative, since the zero case is not really useful IIRC.

Comment (nekevss) [reply src/utils.rs:23]: Temporal Spec has a -1, 0, 1 return

https://tc39.es/proposal-temporal/#sec-temporal-durationsign

File: src/components/duration.rs
Comment (nekevss) [reply src/components/duration.rs]: From what I can tell they both essentially resolve to the same options but with different defaults and methodology. Updated them to a public `DifferenceSettings` and `RoundingOptions` that resolve to `ResolvedRoundingOptions`, let me know what you think!

File: src/components/duration.rs
Comment (jedel1043) [inline src/components/duration.rs:436]: I would make this take the `RoundingOptions` directly, because in the case where someone wants to set an option, they would need to double-wrap with `Some`, which is not ideal.

File: src/options.rs
Comment (jedel1043) [inline src/options.rs:60]: Praise: Loving this new API! I think it really makes sense to have an options struct that resolves all the `Option`s in `RoundingOptions`, instead of having to `unwrap` everywhere.

File: src/components/duration.rs
Comment (jedel1043) [inline src/components/duration.rs:42]: Hmmm, I don't know if it makes much sense to have a conversion from these enums. `Ordering` is a completely different concept than `DurationSign`, and things like `DurationSign::from(zero.cmp(&target))` don't make much sense if you think about it.

File: src/components/duration/time.rs
Comment (nekevss) [reply src/components/duration/time.rs:374]: I went to remove it, but there's still a dependency on the original round. I was going to follow up on this in a different PR.

File: src/components/duration.rs
Comment (nekevss) [reply src/components/duration.rs:42]: Yeah, it's a bit awkward. I wasn't super in love with it. Maybe just a `From<i8>`

File: src/components/duration.rs
Comment (nekevss) [reply src/components/duration.rs]: I'm going back and forth with `DurationSign` vs. `Sign`. :confused: Updated to `Sign` this time around.

File: src/options.rs
Comment (nekevss) [reply src/options.rs]: Updated to a `Sign` value.

File: src/components/duration/normalized.rs
Comment (jedel1043) [inline src/components/duration/normalized.rs:182]: Same suggestion of using the full name.

File: src/lib.rs
Comment (jedel1043) [inline src/lib.rs:110]: If this is to multiply a duration with the sign, what about having an `as_sign_multiplier` function instead that returns `1i8` for zero and positive, and `-1i8` for negative?

Comment (nekevss) [reply src/lib.rs]: It is. It's specifically a nonzero variant in these rounding methods vs. the regular multiplication that occurs elsewhere. I'm fine with `as_sign_multiplier` though. Better than anything I'm coming up with in this instance.

**Review by jedel1043 — APPROVED** [review]
Really impressive work! Pretty much perfect IMO :)

**Review by jasonwilliams — APPROVED** [review]
LGTM!
As @jedel said this looks great, big improvement

**Pull request merged**
